transports/webrtc/src/tokio/sdp.rs

```diff
@@ -0,0 +1,243 @@
+// Copyright 2022 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use serde::Serialize;
+use tinytemplate::TinyTemplate;
+use webrtc::peer_connection::sdp::session_description::RTCSessionDescription;
+
+use std::net::{IpAddr, SocketAddr};
+
+use crate::tokio::fingerprint::Fingerprint;
+
+/// Creates the SDP answer used by the client.
+pub fn answer(addr: SocketAddr, server_fingerprint: &Fingerprint) -> RTCSessionDescription {
+    RTCSessionDescription::answer(render_description(
+        SERVER_SESSION_DESCRIPTION,
+        addr,
+        server_fingerprint,
+        &server_fingerprint.to_ufrag(),
```

Comment:
sorry, maybe you got confused by the comments. We do generate random ufrag in https://github.com/melekes/rust-libp2p/blob/532b41025924b7e6551e21ceb566df0073c8d0e3/transports/webrtc/src/tokio/upgrade.rs#L120, when creating an outbound connection https://github.com/melekes/rust-libp2p/blob/532b41025924b7e6551e21ceb566df0073c8d0e3/transports/webrtc/src/tokio/upgrade.rs#L45

Reply:
> But according to the specification, the server should use that same randomly generated ufrag received from the client when handling the inbound connection.

And it does ->

https://github.com/libp2p/rust-libp2p/blob/7d9d53139ec553ffe2f58d9a0fdff7beffc6c1ec/transports/webrtc/src/tokio/upgrade.rs#L94

There we generate client's offer using the remote ufrag from the incoming ICE request.


`new_inbound_connection` is using the server's fingerprint because it will be used by `peer_connection.create_answer(None)` later when creating answer.